File: Algorithms/708. Insert into a Sorted Circular Linked List/python/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next
"""


class Solution:
    def insert(self, head: "Optional[Node]", insertVal: int) -> "Node":

        if head is None:
            newNode = Node(insertVal, None)
            newNode.next = newNode
            return newNode

        prev, curr = head, head.next
        while curr != head:
            if prev.val <= insertVal <= curr.val:
                # Case #1.
                break
            if prev.val > curr.val:
                # Case #2. where we locate the tail element
                # 'prev' points to the tail, i.e. the largest element!
                if insertVal >= prev.val or insertVal <= curr.val:
                    break
            prev, curr = curr, curr.next
        # Case #3.
        # did not insert the node in the loop
        prev.next = Node(insertVal, curr)
        return head

File: Algorithms/708. Insert into a Sorted Circular Linked List/swift/solution.swift
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var next: Node?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func insert(_ head: Node?, _ insertVal: Int) -> Node? {
        if head == nil {
            let newNode: Node? = Node(insertVal)
            newNode.next = newNode
            return newNode
        }

        var prev: Node? = head
        var curr: Node? = head?.next
        while curr != head {
            if prev!.val <= insertVal && insertVal <= curr!.val {
                break
            }
            if prev!.val > curr!.val {
                if insertVal >= prev!.val || insertVal <= curr!.val {
                    break
                }
            }
            prev = curr
            curr = curr?.next
        }

        var newNode: Node? = Node(insertVal)
        newNode.next = curr
        prev?.next = newNode
        return head
    }
}